tests/regression/75-invalid_dealloc/07-free-at-struct-offset.c

```diff
@@ -0,0 +1,15 @@
+#include <stdlib.h>
+
+typedef struct custom_t {
+    char *x;
+    int y;
+} custom_t;
+
+int main(int argc, char const *argv[]) {
+    custom_t *struct_ptr = malloc(sizeof(custom_t));
+    struct_ptr->x = malloc(10 * sizeof(char));
+    free(struct_ptr->x); //NOWARN
+    free(struct_ptr->y); //WARN
```

Comment:
I suspect this isn't checking the right thing. This is freeing the unknown integer in the field (which also probably comes with some unknown address warning) but it's not checking whether the freeing is of the struct field's address. The latter would be `free(&struct_ptr->y)` I think, which should also warn.

A more interesting case is `free(&struct_ptr->x)`. I suspect that might currently warn, although it shouldn't, because the first field address is the same as the entire struct. It's still a weird thing to do (and we could warn about it), but the freeing would be fine.
So more precise handling might even be to use `Mval.Exp.cmp_zero_offset` to check the offset.

Reply:
> A more interesting case is free(&struct_ptr->x). I suspect that might currently warn, although it shouldn't, because the first field address is the same as the entire struct. It's still a weird thing to do (and we could warn about it), but the freeing would be fine.

It does indeed perform a warning for CWE 761, although it should be totally fine in this case, because it was something that was `malloc`'d and should be `free`'d

Reply:
> So more precise handling might even be to use Mval.Exp.cmp_zero_offset to check the offset.

I'll check it out

Reply:
I used `cmp_zero_offset` in https://github.com/goblint/analyzer/pull/1139/commits/92c0dcae14badbb2be23a62d7c5e15951f877bc2 and it seems to improve the precision.

I see an issue with a false positive warning now, however. More specifically, for the following case::
```c
typedef struct custom_t {
    int y;
    char *y;
} custom_t;
```

If we call `free` on the struct's first field -> no warning happens (so it's even unsound as it seems in this case)
If we call `free` on the struct's second field (provided that it was dynamically allocated) -> a warning pops up (a false positive)

I guess we should try to recursively check for invalid deallocations in order to cover such cases. What do you think?

Reply:
It's not clear to me why recursive checking should be the right thing. In C freeing the outer structure doesn't recursively free other dynamic memory that's pointed to by some field. That needs to be explicitly and separately deallocated.

Reply:
Is there a test case where the improved precision from `cmp_zero_offset` can be seen? If not, then that could be added.

Reply:
> Is there a test case where the improved precision from cmp_zero_offset can be seen? If not, then that could be added.

Yep, both `75/07` and `75/08` were warning for both struct fields before the `cmp_zero_offset` changes. After the `cmp_zero_offset` changes, they both warn only for the 2nd struct field, as expected.